crates/typst/src/layout/grid/layout.rs

```diff
@@ -347,23 +82,225 @@ impl<T: FromValue> FromValue for Celled<T> {
+
+        // Number of content rows: At least as many as given, but also at least
+        // as many as needed to place each item.
+        let r = {
+            let len = cells.len();
+            let given = tracks.y.len();
+            let needed = len / c + (len % c).clamp(0, 1);
+            given.max(needed)
+        };
+
+        let has_gutter = gutter.any(|tracks| !tracks.is_empty());
+        let auto = Sizing::Auto;
+        let zero = Sizing::Rel(Rel::zero());
+        let get_or = |tracks: &[_], idx, default| {
+            tracks.get(idx).or(tracks.last()).copied().unwrap_or(default)
+        };
+
+        // Collect content and gutter columns.
+        for x in 0..c {
+            cols.push(get_or(tracks.x, x, auto));
+            if has_gutter {
+                cols.push(get_or(gutter.x, x, zero));
+            }
+        }
+
+        // Collect content and gutter rows.
+        for y in 0..r {
+            rows.push(get_or(tracks.y, y, auto));
+            if has_gutter {
+                rows.push(get_or(gutter.y, y, zero));
+            }
+        }
+
+        // Remove superfluous gutter tracks.
+        if has_gutter {
+            cols.pop();
+            rows.pop();
+        }
+
+        // Reverse for RTL.
+        let is_rtl = TextElem::dir_in(styles) == Dir::RTL;
+        if is_rtl {
+            cols.reverse();
+        }
+
+        Self { cols, rows, cells, has_gutter, is_rtl }
+    }
+
+    /// Resolves all cells in the grid before creating it.
+    /// Allows them to keep track of their final properties and adjust their fields accordingly.
+    #[allow(clippy::too_many_arguments)]
+    pub fn resolve<T: ResolvableCell>(
+        tracks: Axes<&[Sizing]>,
+        gutter: Axes<&[Sizing]>,
+        cells: Vec<T>,
+        fill: &Celled<Option<Paint>>,
+        align: &Celled<Smart<Align>>,
+        inset: Sides<Rel<Length>>,
+        engine: &mut Engine,
+        styles: StyleChain,
+    ) -> SourceResult<Self> {
+        // Number of content columns: Always at least one.
+        let c = tracks.x.len().max(1);
+
+        let mut cells = cells
+            .into_iter()
+            .enumerate()
+            .map(|(i, cell)| {
+                let x = i % c;
+                let y = i / c;
+
+                Ok(cell.resolve_cell(
+                    x,
+                    y,
+                    &fill.resolve(engine, x, y)?,
+                    align.resolve(engine, x, y)?,
+                    inset,
+                    styles,
+                ))
+            })
+            .collect::<SourceResult<Vec<_>>>()?;
+
+        // If not all columns in the last row have cells, we will add empty
+        // cells and complete the row so that those positions are susceptible
+        // to show rules and receive grid styling.
+        let cell_count = cells.len();
+        if cell_count % c != 0 {
+            let cells_remaining = c - (cell_count % c);
+            cells.reserve_exact(cells_remaining);
```

Comment:
Instead of this reserve, we could also allocate the full cell vector with_capacity upfront and then use a for loop to populate it above. This way we can also skip the if condition here, I think.

Reply:
Do you mean calculating how many cells will be needed with something like `let capacity = cells.len() + cells.len() % c`, and then doing a `for i in 0..capacity` instead of looping on `cells` itself, taking `cells.get(i).unwrap_or_else(T::default)`, resolving it and pushing it to the final vector on each iteration?

That could perhaps work. I'll have to allocate a new vector with `with_capacity` anyway in my `x`,`y` fields PR (since cells will be potentially moved around), though I think I'll have to keep this if with reserve in the next PR, since we can't predict how long the final cells vector will be when cells can have arbitrary positions (we can only predict its minimum length). For this PR, though, this can probably work. (Not sure how it behaves in terms of performance, but hopefully shouldn't be too bad?)

Reply:
I meant the capacity calculation like you said. I didn't necessarily mean putting everything into a single loop, but that might work as well! I was just thinking that if we reserve everything up-front, then the cells_remaining loop doesn't need to be in an if because it would just run zero times. I'll leave up to you whether a single loop or two loops work better.

Reply:
Oooh!

Actually, initially the `if` wasn't there, but a lot of tests started failing... Turns out that, if `cell_count % c == 0` (the row is already filled), `cells_remaining` would be equal to `c - 0`, a.k.a. `c`, so it would add more cells than necessary. I couldn't find a better alternative at the time... though, using early allocation as you suggest, we could probably just use `let cells_remaining = cells.capacity() - cells.len()` instead :p

Reply:
Ah, I see. That makes sense.

Reply:
From https://doc.rust-lang.org/std/vec/struct.Vec.html#method.with_capacity:

> The vector will be able to hold at least capacity elements without reallocating.

I don't think we can rely on calling cells.capacity(). But we can use the variable we called with_capacity with.

Reply:
I've simplified the overall structure of the code. I couldn't manage to remove the `if`, unfortunately.

I tried to use `Vec::with_capacity`, but I would necessarily have to change things to use a `for` loop with `push(next())` or something - I thought of using `.extend()` to avoid doing that, but turns out the usage of `SourceResult` inside iterators is a pain (I'd have to `collect` and end up creating yet another vector!). I will have to eventually create that for loop and stuff, but I thought the new code is simpler for now; let me know what you think.

Reply:
Regarding possible capacity optimizations, I believe the iterator holds the correct size hint all throughout the iterator "pipeline" (see here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b565d7bb84801927b1029389d1e305de ), so we should probably be fine I think, assuming `collect` is smart enough.